8290154: [JVMCI] partially implement JVMCI for RISC-V

Zeavee opened the pull request:
This patch adds a partial JVMCI implementation for RISC-V, to allow using the GraalVM Native Image RISC-V LLVM backend, which does not use JVMCI for code emission.
It creates the jdk.vm.ci.riscv64 and jdk.vm.ci.hotspot.riscv64 packages, as well as implements a part of jvmciCodeInstaller_riscv64.cpp. To check for correctness, it enables JVMCI code installation tests on RISC-V. More testing is performed in Native Image.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8290154](https://bugs.openjdk.org/browse/JDK-8290154): [JVMCI] partially implement JVMCI for RISC-V


### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [8742b9b2](https://git.openjdk.org/jdk/pull/9587/files/8742b9b27a3467e7ce11f082cdf68f3244479332)
 * [Doug Simon](https://openjdk.org/census#dnsimon) (@dougxc - Committer) ⚠️ Review applies to [c3c6f651](https://git.openjdk.org/jdk/pull/9587/files/c3c6f651f52418eedfaed8971f4cae1c53da89fd)
 * [Yadong Wang](https://openjdk.org/census#yadongwang) (@yadongw - Author) ⚠️ Review applies to [225fee42](https://git.openjdk.org/jdk/pull/9587/files/225fee42a1892aa4c42572287d19e3b6315abf7a)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9587/head:pull/9587` \
`$ git checkout pull/9587`

Update a local copy of the PR: \
`$ git checkout pull/9587` \
`$ git pull https://git.openjdk.org/jdk pull/9587/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9587`

View PR using the GUI difftool: \
`$ git pr show -t 9587`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9587.diff">https://git.openjdk.org/jdk/pull/9587.diff</a>

</details>

Zeavee commented:
/covered

RealFYang commented:
Hi, I see some JVM crash when I try the following test with fastdebug build with your patch:
    test/hotspot/jtreg/compiler/jvmci/jdk.vm.ci.code.test/src/jdk/vm/ci/code/test/SimpleDebugInfoTest.java

# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/fyang/openjdk-jdk/src/hotspot/cpu/riscv/nativeInst_riscv.cpp:118), pid=1154    063, tid=1154084
#  assert(NativeCall::is_call_at((address)this)) failed: unexpected code at call site

Zeavee commented:
> Hi, I see some JVM crash when I try the following test with fastdebug build with your patch: test/hotspot/jtreg/compiler/jvmci/jdk.vm.ci.code.test/src/jdk/vm/ci/code/test/SimpleDebugInfoTest.java
Thank you for pointing this out, I did not run the patch with fastdebug. I saw other issues after solving this one, so I take some time to solve them as well.

turbanoff commented on src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.riscv64/src/jdk/vm/ci/riscv64/RISCV64Kind.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    public boolean isFP() {
+        switch(this) {
```
let's add space after `switch`

Zeavee replied:
Sure!

magicus reviewed (approved):
Build changes look ok. You still need reviews for the actual code changes.

magicus commented:
/reviewers 3

dougxc commented:
Please change the title of this PR and the JBS issue to "[JVMCI] partially implement JVMCI for RISC-V" to more accurately reflect the code changes.

Zeavee commented:
Sorry for the delay, but all the JVMCI now pass.

dougxc commented:
@vnkozlov can you please help review this PR.

RealFYang commented on src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.riscv64/src/jdk/vm/ci/riscv64/RISCV64Kind.java:
```diff
@@ -0,0 +1,171 @@
+    V256_DWORD(32, DWORD),
+    V256_QWORD(32, QWORD),
+    V256_SINGLE(32, SINGLE),
+    V256_DOUBLE(32, DOUBLE);
```
I see some vector types are added here. Could you please explain the design here?
Please note that RISC-V Vector extension (RVV) uses variable length vectors which in this sense is similar to AArch64's SVE. The actual vector length depends on the hardware implementation and could be determind by reading from the corresponding CSR special register at runtime. I am not sure how JVMCI is supposed to support this.

Zeavee replied:
At the moment, only AArch64's Neon is supported by JVMCI, which means there is no other example for variable length vector architectures. I think that a correct implementation would indeed fetch the value of the corresponding CSR at runtime. This would mainly be needed in the `getLargestStorableKind` method, as it depends on the hardware. I imagine the values of those registers should also be accessible through the `HotSpotVMConfig`, to allow the compilers to know the hardware implementation.

In this case, the implementation is only used in Native Image, which currently does not implement the vector extension for RISC-V. Thus, I chose to manually define valid vector types, but since they can have a size up to 65536, I thought it would not be a good idea to define all of them.

Since this is not needed in the current Native Image LLVM backend for RISC-V, I'm not sure I would have the time to work on this, so if it would be better not to have those "placeholder" types, I can remove them for now.

RealFYang replied:
Thanks for the explaination. Then I would suggest we remove those vector types together with the SIMD register definitions for now.

RealFYang reviewed (commented):
I have some comments.

RealFYang commented:
> Sorry for the delay, but all the JVMCI tests now pass.

So I give it a second try with fastdebug build on my unmatched board [1], I see the error has changed into [2].
Is there anything I missed? Thanks.

[1]
$ make run-test-only TEST="./test/hotspot/jtreg/compiler/jvmci/jdk.vm.ci.code.test/src/jdk/vm/ci/code/test/SimpleDebugInfoTest.java" JTREG="TIMEOUT_FACTOR=16"

[2]
JavaTest Message: JUnit Failure: testRegNarrowObject(jdk.vm.ci.code.test.SimpleDebugInfoTest): null
java.lang.AssertionError
        at jdk.vm.ci.code.test.riscv64.RISCV64TestAssembler.f(RISCV64TestAssembler.java:62)
        at jdk.vm.ci.code.test.riscv64.RISCV64TestAssembler.emitLui(RISCV64TestAssembler.java:127)
        at jdk.vm.ci.code.test.riscv64.RISCV64TestAssembler.emitLoadImmediate(RISCV64TestAssembler.java:140)
        at jdk.vm.ci.code.test.riscv64.RISCV64TestAssembler.emitLoadPointer32(RISCV64TestAssembler.java:300)
        at jdk.vm.ci.code.test.riscv64.RISCV64TestAssembler.emitLoadPointer(RISCV64TestAssembler.java:317)
        at jdk.vm.ci.code.test.SimpleDebugInfoTest.lambda$testRegNarrowObject$12(SimpleDebugInfoTest.java:262)
        at jdk.vm.ci.code.test.DebugInfoTest.lambda$test$0(DebugInfoTest.java:68)
        at jdk.vm.ci.code.test.CodeInstallationTest.test(CodeInstallationTest.java:104)
        at jdk.vm.ci.code.test.DebugInfoTest.test(DebugInfoTest.java:61)
        at jdk.vm.ci.code.test.DebugInfoTest.test(DebugInfoTest.java:52)
        at jdk.vm.ci.code.test.SimpleDebugInfoTest.testObjectOnStack(SimpleDebugInfoTest.java:209)
        at jdk.vm.ci.code.test.SimpleDebugInfoTest.testRegNarrowObject(SimpleDebugInfoTest.java:266)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
        at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:105)
        at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:62)
        at org.junit.runner.JUnitCore.runClasses(JUnitCore.java:49)
        at com.sun.javatest.regtest.agent.JUnitRunner.main(JUnitRunner.java:76)
        at com.sun.javatest.regtest.agent.JUnitRunner.main(JUnitRunner.java:43)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
        at java.base/java.lang.Thread.run(Thread.java:1589)

java.lang.Exception: JUnit test failure
        at com.sun.javatest.regtest.agent.JUnitRunner.main(JUnitRunner.java:92)
        at com.sun.javatest.regtest.agent.JUnitRunner.main(JUnitRunner.java:43)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
        at java.base/java.lang.Thread.run(Thread.java:1589)

JavaTest Message: Test threw exception: java.lang.Exception: JUnit test failure
JavaTest Message: shutting down test

Zeavee commented:
> So I give it a second try with fastdebug build on my unmatched board [1], I see the error has changed into [2]. Is there anything I missed? Thanks.

I don't know why I did not get this assertion error on my end before. I fixed the issue and rebuilt the project to check it passes. I hope there is nothing else about this.

Also, I removed the Vector registers and Kinds.

dougxc commented on src/hotspot/share/runtime/deoptimization.hpp:
```diff
@@ -342,7 +342,7 @@ class Deoptimization : AllStatic {
       return 0;
     }
   }
-  static int trap_request_index(int trap_request) {
```
Please add a comment as to why `noinline` is required here.

Zeavee commented:
Do you have details about testing performed in Native Image as mentioned in PR decription?

Yes, the RISC-V LLVM backend for Native Image passes 99% of the tests performed, which is similar to the other LLVM backends.

> I see you added more changes in hotspot file sharedRuntime_riscv.cpp guarded by macro INCLUDE_JVMCI. Searching for INCLUDE_JVMCI or COMPILER2_OR_JVMCI in src/hotspot/cpu/aarch64, I see several more places checking for these macros. Have you checked if we need similar changes for your use case?

I first added the changes for all places where those macros are used, but since only modifying sharedRuntime_riscv.cpp was enough to make the tests pass, I did not wanted to add code that I was not sure was useful at the moment.

> Also could you explain the change made in hotspot file deoptimization.hpp? Thanks.

When the method is inlined, the `if (trap_request < 0)` check behaves incorrectly when the `nmethod` is compiled by JVMCI. Even though the boolean is true, the function returns -11 instead of -1, and the `if (unloaded_class_index >= 0)` checks have the same issue, causing an access to an illegal index of an array. I am not sure why this happens, as it works correctly for method not compiled by JVMCI.

yadongw commented on src/hotspot/cpu/riscv/jvmciCodeInstaller_riscv.cpp:
```diff
@@ -0,0 +1,128 @@
+      return as_FloatRegister(floatRegisterNumber)->as_VMReg();
+    }
+    jint vectorRegisterNumber = floatRegisterNumber - FloatRegisterImpl::number_of_registers;
+    if (vectorRegisterNumber >= 0 && vectorRegisterNumber - VectorRegisterImpl::number_of_registers) {
```
Maybe you want to check if `vectorRegisterNumber < VectorRegisterImpl::number_of_registers`.

Zeavee replied:
Indeed, it is incorrect. In fact, since I removed the vector registers for now, I should just remove this part.

yadongw commented on src/hotspot/share/runtime/deoptimization.hpp:
```diff
@@ -342,7 +342,10 @@ class Deoptimization : AllStatic {
+  // In RISC-V, when this method is inlined, the condition behaves
+  // incorrectly if the nmethod is compiled by JVMCI, causing the
+  // method to return a negative index which is not -1
+  static int RISCV64_ONLY(__attribute__ ((noinline))) trap_request_index(int trap_request) {
```
I don't understand this change, because the semantics should be the same whether inline or not. What's the difference  in the generated code by the c/c++ compiler?

Zeavee replied:
Here is the assembly code without inlining:
```
  49b768:	57fd                	li	a5,-1
  49b76a:	00054363          	bltz	a0,49b770 <_ZN14Deoptimization18trap_request_indexEi+0xe>
  49b76e:	87aa                	mv	a5,a0
  49b770:	6422                	ld	s0,8(sp)
  49b772:	0007851b          	sext.w	a0,a5
```

and here is the code with inlining:
```
  7de6a4:	000a879b          	sext.w	a5,s5
  7de6a8:	000ad363          	bgez	s5,7de6ae <_ZN14Deoptimization19uncommon_trap_innerEP10JavaThreadi+0x2ea>
  7de6ac:	57fd                	li	a5,-1
  7de6ae:	757d                	lui	a0,0xfffff
```
(a1 was moved in s5).
I don't see any semantic difference, so I am not sure why it would change. However, I tried to build and test both versions several times and the issue only appeared when the method was inlined.

Zeavee replied:
Sorry for the delay, but I managed to remove the attribute. There was a value that was not sign extended, but it had to be despite being an int. For some reason, it seems that removing the inlining sign extended the value.

RealFYang commented:
> > I see you added more changes in hotspot file sharedRuntime_riscv.cpp guarded by macro INCLUDE_JVMCI. Searching for INCLUDE_JVMCI or COMPILER2_OR_JVMCI in src/hotspot/cpu/aarch64, I see several more places checking for these macros. Have you checked if we need similar changes for your use case?
> 
> I first added the changes for all places where those macros are used, but since only modifying sharedRuntime_riscv.cpp was enough to make the tests pass, I did not wanted to add code that I was not sure was useful at the moment.

Well, that sounds fragile to me since you are depending on a relatively small set of JTreg tests here. I think an analysis is needed here to be sure about whether those are really needed or not.

Zeavee commented:
> > > I see you added more changes in hotspot file sharedRuntime_riscv.cpp guarded by macro INCLUDE_JVMCI. Searching for INCLUDE_JVMCI or COMPILER2_OR_JVMCI in src/hotspot/cpu/aarch64, I see several more places checking for these macros. Have you checked if we need similar changes for your use case?
> > 
> > 
> > I first added the changes for all places where those macros are used, but since only modifying sharedRuntime_riscv.cpp was enough to make the tests pass, I did not wanted to add code that I was not sure was useful at the moment.
> 
> Well, that sounds fragile to me since you are depending on a relatively small set of JTreg tests here. I think an analysis is needed here to be sure about whether those are really needed or not.

Those parts of code will ultimately be needed for a full JVMCI implementation, but adding them right now would leave them untested, since for example, neither of the JTreg tests or the RISC-V LLVM backend for Native Image cover the case where synchronized code is installed (this corresponds to the code in `templateInterpreterGenerator_riscv.cpp`).

However, if it is still better to include those parts of code, I don't mind adding the implementation I have for them.

Zeavee commented:
Hello, this PR has been stuck for some time now. What should I do to proceed?

dougxc commented:
@RealFYang are you ok with deferring further changes to a future RFE?

RealFYang commented:
> Hello, this PR has been stuck for some time now. What should I do to proceed?

The current version does not build. I will take another look after this is rebased on the latest jdk master.

Zeavee commented:
> The current version does not build. I will take another look after this is rebased on the latest jdk master.

Sorry for the delay, I rebased the PR and fixed the building issue.

yadongw reviewed (approved):
lgtm(not a reviewer)

Zeavee commented:
/integrate

dougxc commented:
/sponsor
